### ISSUE TITLE
No 'Access-Control-Allow-Origin' header is present on the requested resource.

### ISSUE THREAD

**Issue by Steague** [okta/okta-oidc-js#337]
## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [x] okta-angular
- [ ] oidc-middleware
- [X] okta-react
- [ ] okta-react-native
- [x] okta-vue

## I'm submitting a:

- [X] Bug report  
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior

When logging in, Okta redirects the user back to [url]/implicit/callback with the token info attached to the URI. When that happens, the user is the redirected again to /implicit/callback without the token and `AuthApiError` is shown.

## Expected behavior

Login works as normal and the token is stored correctly in localStorage.

## Minimal reproduction of the problem with instructions

I am unable to provide a demo due to this only happening after being deployed. (I cannot reproduce the issue locally even if I build the app locally)

## Extra information about the use case/user story you are trying to implement

My local development toolchain consists of two node instances running. The first instance is the UI (which contains the okta-react package). The second Node instance is the backend (services stack).

The UI is based in Create-React-App and is built using react-scripts (Webpack derivative)
The backend is plain old Node/express that is packed with traditional Webpack.

Running okta-react@1.0.2 works as expected but has issues regarding paths within the UI router that have been fixed by 1.1.3. The only thing changed is okta-react to v1.1.3 (and any deps that it needs) and then the error happens.

## Environment

- Package Version: 1.1.3
- Browser: Chrome
- OS: MacOS and Windows
- Node version (`node -v`): 8.11.2
- Other:

**Comment by rodoabad** [comment]
@Steague this usually happens when you have different domains i.e. you logged in at localhost earlier and now you are trying to login to a different domain even though they are added to the same Okta application.

I'm not sure if this is intended or not but it's something that's a little bit painful when you're developing in `localhost` and you want to check in a deployed `dev` environment.

**Comment by Steague** [comment]
@rodoabad Thanks for your reply. I'm getting the error after I clear my entire cache. I also get the error when in an incognito window, a different browser than the one I'm developing in, and on a different physical machine than the one I'm developing on.

Also, the behavior does not occur when using 1.0.2 of okta-react, only 1.1.3.

**Comment by bartimaeus** [comment]
I've run into the same issue when updating dependencies. I ended up locking @okta/okta-react to 1.0.2 and everything is working again.

This same issue is also mentioned in [okta-signin-widget issue #541](https://github.com/okta/okta-signin-widget/issues/541)

**Comment by bartimaeus** [comment]
Nevermind, it is still not working. I cleared my browser cache and tried both my development and staging environments with previous working dependencies and there was a failure on the second authentication because the first authentication appears to be cached somewhere on the Okta end.

Both domains are setup for CORS. What is the suggestion for this use case? Should we use separate Okta accounts for each environment rather than creating apps inside of Okta for development and staging?

**Comment by Steague** [comment]
@bartimaeus Using  v1.0.2 is not an optimal option for me as the `exact` property is not correctly working. In the end, I had to use v1.0.2 and then also setup my app router like this:

```
    render() {
        return (
            <Router>
                <Security
                    issuer={`https://${this.okta.host}/oauth2/${this.okta.serverId}/`}
                    client_id={`${this.okta.clientId}`}
                    redirect_uri={window.location.origin + '/implicit/callback'}
                >
                    {(window.location.href).indexOf('unauthorized') === -1 ?
                        <SecureRoute path='/' exact={true} component={Main} /> :
                        <Route path='/unauthorized' exact={true} component={Unauthorized} />
                    }
                    <Route path='/implicit/callback' exact={true} component={ImplicitCallback} />
                </Security>
            </Router>
        );
    }
```

**Comment by bartimaeus** [comment]
@Steague that is interesting that your `exact` property isn't working correctly. I  am not running into that issue.

I was able to track down my issue. I had updated **@okta/okta-auth-js** to `2.0.1` however **@okta/okta-signin-widget** depends on **@okta/okta-auth-js:1.17.0**. After reverting my okta-auth-js commit everything worked again as expected. The other issue is that **@okta/okta-react:1.1.1+** updated the **@okta/okta-auth-js** dependency to `2.0.1` as well.

My final locked versions are:

```
"@okta/okta-auth-js": "^1.17.0",
"@okta/okta-react": "^1.0.2",
"@okta/okta-signin-widget": "^2.13.0",
```

My routes are setup as follows:

```
<Fragment>
  <UpdateNotification />
  <Provider store={store}>
    <Router>
      <Security
        issuer={issuer}
        client_id={clientId}
        redirect_uri={redirectUri}
        onAuthRequired={onAuthRequired}
      >
        {process.env.REACT_APP_GOOGLE_ANALYTICS_ID && (
          <Route component={GoogleAnalytics} />
        )}
        <Switch>
          <SecureRoute exact path="/" component={AuthenticatedLayout} />
          <Route path="/login" component={PublicLayout} />
          <Route path="/register" component={PublicLayout} />
          <Route path="/implicit/callback" component={ImplicitCallback} />
          <SecureRoute path="/*" component={AuthenticatedLayout} />
        </Switch>
      </Security>
    </Router>
  </Provider>
</Fragment>
```

**Comment by Steague** [comment]
@bartimaeus I'm currently on 

`"@okta/okta-react": "1.0.2"`

I am using Create-React-App and when I attempted to migrate to `v1.1.3` then it also upgraded the okta-auth-js package from

`"@okta/okta-auth-js": "^1.14.0"`

to

`"@okta/okta-auth-js": "^1.17.0"`

which errored out when attempting to transpile with the version of the react-scripts package that came with CRA so I had to update react-scripts to:

`"react-scripts": "^2.0.5"`

That then allowed me to transpile, but that also caused the error I got that began this open ticket. I have now reverted back to `v1.0.2` since i don't get the error I originally posted, the code transpiles, but the `exact` prop doesn't work in my router.

I have tried MANY nominations of the above packages but have only got my hacky workaround to get the app to work in all my use cases.

**Comment by bartimaeus** [comment]
I hear you. I am using CRA with `react-scripts:1.1.4` and `react-app-rewired:1.6.2`.

**Comment by JonoParry** [comment]
I am also having this issue with okta-angular.

2 sites deployed, prod and staging, which have 2 separate registered client apps.

> Access to XMLHttpRequest at 'https://**********okta.com/oauth2/default/v1/keys' from origin '{PROD URL}' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header has a value '{STAGING URL}' that is not equal to the supplied origin.

Followed by the `AuthApiError`

Empty cache and hard reload temporarily fixes the issue.

**Comment by shravan-kaushik** [comment]
I have the exact same issue as @JonoParry , but with okta-vue

**Comment by dotnetcoder01** [comment]
> I am also having this issue with okta-angular.
> 
> 2 sites deployed, prod and staging, which have 2 separate registered client apps.
> 
> > Access to XMLHttpRequest at 'https://**********okta.com/oauth2/default/v1/keys' from origin '{PROD URL}' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header has a value '{STAGING URL}' that is not equal to the supplied origin.
> 
> Followed by the `AuthApiError`
> 
> Empty cache and hard reload temporarily fixes the issue.

I have this same issue:
Okta Angular, multiple separate apps registered, cors errors with AuthApiError  being thrown.

**Comment by cerikpete** [comment]
Just wanted to add that I also had to back okta-react down to 1.0.2 (from 1.1.4) to get it to work.

**Comment by jmelberg-okta** [comment]
Hi all,

We've been tracking this API bug [in another repository](https://github.com/okta/okta-signin-widget/issues/541).

**Good news** - the fix for all Okta organizations will be available next week. More information is available in [this comment](https://github.com/okta/okta-signin-widget/issues/541#issuecomment-446411683).

I'll keep this issue open until this fix has been rolled out to all Okta organizations.

Thank you all for your patience.

**Comment by Steague** [comment]
@jmelberg-okta I see in your reply in the other repo that you've rolled out a fix to preview org (*.oktapreview.com). I'm not sure if that includes me but I use a *.oktapreview.com domain for my project. Will the fix be an updated `okta-react` package? Updated `okta-oidc-js` package? Both? What will I need to do to update my app? Also, should I be seeing anything different happening now since the merged fix is rolled out to oktapreview domains?

**Comment by Steague** [comment]
@jmelberg-okta Is my new ticket also related to thefix you speak to? https://github.com/okta/okta-oidc-js/issues/354

**Comment by jmelberg-okta** [comment]
@Steague,

Yep - if you're using a `*.oktapreview.com` domain, the issue you described in #354 will most likely be resolved.

In previous versions of this SDK (`okta-react@1.0.2 and below`), we were pinned to a version of our authentication SDK [**okta-auth-js**](https://github.com/okta/okta-auth-js/blob/master/CHANGELOG.md), that was requesting the `/oauth2/{authServerId}/v1/keys` endpoint on demand. This behavior hid the API bug from us, as it would only be encountered in specific scenarios.

By updating to the latest **okta-react** version, you're using an updated version of **okta-auth-js**, which makes that request by default when an `idToken` is received. Hence, bringing to light the API issue.

If issues persist, please let us know!

**Comment by karsai5** [comment]
Heya, I believe this is related (if not let me know to create a new ticket). The testing versions of our app are having issues logging in and I think I narrowed it down to the following change. When `okta-react` makes a request to *.oktapreview.com/oauth2/v1/keys it returns with the following header: `Access-Control-Allow-Origin: *`. I checked in the prod version of our app; and the same request returns with the url of the host that made the request in the `Access-Control-Allow-Origin` header. 

Because of this the request is rejected by the browser with the following error: `has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.`

**Comment by nfrechette** [comment]
@karsai5 I am having the same issue with the okta-vue package. It is very strange, two days ago it was working and today it isn't and AFAIK nothing really changed on my end, no packages were updated, just webserver reboot.

**Comment by rlarasarmiento** [comment]
@jmelberg-okta I'm having the same issue of @nfrechette on my end. Five hours ago it was working.

**Comment by Jag-Nahl** [comment]
@nfrechette @karsai5 I've got the exact same issue on our development website. Turned off our app service on monday and restarted it today and its failing to login.

**Comment by smbakker** [comment]
@jmelberg-okta I'm having the same issue as @karsai5 in the angular packages. I noticed that it is not an issue on our ***.okta-emea.com tenant. `Access-Control-Allow-Origin` has the right value and our angular app is working. Please fix in the *.oktapreview.com tenants.

**Comment by mattanknol** [comment]
@jmelberg-okta We are also experiencing the issue described by @karsai5 and all the others.
We tested and this issue only exists on the `*.oktapreview.com` domain (our development environment) and not our apps running on `*.okta-emea.com` (our staging and production environments).

**Comment by hermanvantriest** [comment]
@jmelberg-okta 
Because of the update of OKTA the endpoint /oauth2/default/v1/keys returns `Access-Control-Allow-Origin: *` instead of `Access-Control-Allow-Origin: <CORS URL>`. This causes that the latest `@okta/okta-signin-widget` npm package is not working anymore. This causes the following exception when the token is being validated in the oauth implicit callback:
`Access to XMLHttpRequest at 'https://yyy.oktapreview.com/oauth2/default/v1/keys' from origin '<CORS URL>' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.`

Can you please rollback the update or update the Angular packages?

**Comment by noelreardon** [comment]
Exactly same issue here with Angular packages. Please rollback.
Customer demo has just been shot.

**Comment by robertjd** [comment]
Hello all, for those affected a CORS issue this morning the issue should now be resolved, please see https://trust.okta.com/#incident/a9C1Y000000GpJuUAK and open a support ticket via support@okta.com if you continue to see problems with your environment. Thank you!

**Comment by kezek** [comment]
Same issue on okta-vue (1.0.7). Please investigate

**Comment by RickBross** [comment]
> Hello all, for those affected by a CORS issue this morning the issue should now be resolved, please see https://trust.okta.com/#incident/a9C1Y000000GpJuUAK and open a support ticket via [support@okta.com](mailto:support@okta.com) if you continue to see problems with your environment. Thank you!

Still experiencing this issue in the wild. Using...
```
    "@okta/okta-react": "^1.1.4",
    "@okta/okta-auth-js": "^2.2.0",
```

**Comment by nbarbettini** [comment]
@RickBross What's the error message you are seeing?

FYI, I posted an update on a related issue here: https://github.com/okta/okta-signin-widget/issues/541#issuecomment-451044362

**Comment by rk-7** [comment]
Still experiencing this issue. Any solutions yet?

**Comment by zah33r** [comment]
@nbarbettini We're also facing the same issue with multiple domains. Its been embarrassing telling customers to clear cache every time they switch environments. Desperately looking for a solution.

**Comment by WtfJoke** [comment]
@rk-7 and @zah33r Please stop commenting here and write an email to support@okta.com like written in the linked comment from @nbarbettini  https://github.com/okta/okta-signin-widget/issues/541#issuecomment-458766353 to get the fix for your customers.

Im interested if they have succesfully rolled out the feature globally.

**Comment by jagdishhulsure** [comment]
Access to XMLHttpRequest at 'https://ecapisolution.azurewebsites.net/api/sell/data/1/' (redirected from 'http://ecapisolution.azurewebsites.net/api/sell/data/1/') from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource in reactjs, Can anyone help to resolve this issue. thank you in advance.

**Comment by swiftone** [comment]
@jagdishhulsure - Have you double checked that the relevant domains are listed in the Trusted Origins (and CORS checked) in your Okta Admin Dashboard?

**Comment by jagdishhulsure** [comment]
@swiftone not checked, but the same other dummy API's are working "https://reqres.in/api/users", when am using this API 'https://ecapisolution.azurewebsites.net/dummy/api/link' it's throwing an error in console.
this is the same error  Access to XMLHttpRequest at 'https://ecapisolution.azurewebsites.net/dummy/api/link' (redirected from 'http://ecapisolution.azurewebsites.net/dummy/api/link/') from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource

**Comment by swiftone** [comment]
@jagdishhulsure - I'm not entirely sure I follow, but you say one domain is working, but another domain isn't, so I'd recommend actually checking to make sure the domain that is failing is listed as a Trusted Origin with CORS checked.

Let us know if that's not the issue.

**Comment by nbarbettini** [comment]
Cross-posting for visibility: https://github.com/okta/okta-signin-widget/issues/541#issuecomment-507474696

We've rolled out a global fix for these types of errors. If you're still seeing them, please open a separate issue. Thanks!
